### DISCUSSION TITLE
Handle data fetching errors in getStaticProps/getStaticPaths

### DISCUSSION THREAD

**Discussion by balazsorban44** [vercel/next.js#11180] [Help] [answered]
I have the following `getStaticProps` function:

```js
export const getStaticProps = async (ctx) => {
    try {
        const {slug} = ctx.params
        const url = Array.isArray(slug) ? slug.join("/") : slug
        const token = await getToken()
        const result = await getData(url, token)

        if (result.errors) {
            throw result.errors
        }

        if (!result.data) {
            throw {message: "No data"}
        }

        return ({props: result.data})

    } catch (error) {
        error.ctx = ctx
        logErrorToExternalLoggingService(error, "Fetching initial props failed")
        return ({props: {}})
    }
}
```

In addition, in `getStaticPaths`, I set `fallback` to `true`, so as discussed in #11134, `getStaticProps` should first run at runtime, not at build time.

Now if the API fails this first time, the way I handle it now, props will not be populated correctly, and all subsequent requests will receive an empty object. Is there a way to get around this? Somehow tell getStaticProps that if something failed at the first request, the next time it should try again fetching the data, instead of returning the empty object?

Can you recommend any other way to solve this?

If it has any importance, when a request comes, I create an ApolloClient instance to fetch the data using, the `url`.

**Comment by Timer** [comment, marked as answer]
@balazsorban44 you should not attempt to `try`/`catch` any error(s) in your `getStaticProps`. This error should be "unhandled", and Next.js will automatically render a default error page (or your [customized version](https://nextjs.org/docs/advanced-features/custom-error-page#500-page)), where you can read the reason and render something appropriately.

Next time a user visits this page again, Next.js will attempt to render it a second time (and so on).

**Reply by balazsorban44** [reply]
Thanks, I'll check that! I still would like to log the error, but then I just `throw` it further in the catch block, I hope that will work.

**Reply by Timer** [reply]
Yup! Something like this should work, then:
```js
    } catch (error) {
        error.ctx = ctx
        logErrorToExternalLoggingService(error, "Fetching static props failed")
        throw error
    }
```

**Reply by balazsorban44** [reply]
Just to be sure, is the behavior the same in `getStaticPaths`? Will an error while fetching paths be handled by Next.js internally as well, and the user is sent to the error page?

**Reply by timneutkens** [reply]
An error thrown in getStaticPaths will stop the build and exit.

**Reply by balazsorban44** [reply]
Thank you!

**Reply by luisdiaslima** [reply]
@balazsorban44 my build is canceled when errors occur in `getStaticProps`. How do i ignore this?

**Reply by tomgreeEn** [reply]
I have a similar question, but am wondering if there is any way to catch an error further up the chain (while still allowing it to fail ultimately).

My use case is that my getStaticProps function can succeed, but the page build fails as a result of `"Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value.” ` - ie after Next has run further checks and found invalid data.

I am fetching data from a CDN to allow me to build a large number of pages in the time limit, and occasionally stale data can result in missing values which cause the error above. I would like the chance to send an external error and purge the CDN cache for a page if it fails at this point, so that on the next request/build it can get fresh data which may have been fixed in the CMS. Apart from adding my own `is-serializable-props` check before returning getStaticProps, is there a more robust and general way to tap into the props validation process or error handling (without interfering with it ultimately)?

**Reply by bayukp** [reply]
Thank you! Imo this is important and I don't see it in documentation section under getStaticProps implementation (pls correct me)

**Reply by lucaspieran** [reply]
I have done what you say and it doesn't work. My cms is off at a certain time and during that time my pages that depend on the cms throw 500, shouldn't they show the last successful build??? @Timer 

```javascript
export const getStaticProps: GetStaticProps = async ({ locale }) => {
  const res = await fetch(`https://.../api/courses?locale=${locale}&populate=*`)
  if (!res.ok) {
    throw new Error('error')
  }
  const { data } = await res.json()
  return {
    props: {
      courses: data,
      ...(await serverSideTranslations(locale ?? 'en', ['solutions'])),
    },
    revalidate: 60,
  }
}
```
